### ISSUE TITLE
A diagnostics mode to report when constraints aren't being met for optimal performance

### ISSUE THREAD

**Issue by stas00** [pytorch/pytorch#70951]
### 🚀 The feature, motivation and pitch

As its known CUDA has all kinds of constraints for certain optimizations to kick in. Often it's not obvious when such constraints aren't met and a user ends up with pretty bad performance and they are not the wiser.

---------------

Let's start with a use-case of **a very not obvious** problem. 

I've done multiple benchmarks for [A100](https://github.com/huggingface/transformers/issues/15026) and [RTX-3090](https://github.com/huggingface/transformers/issues/14608). One of these was to compare fp16 vs bf16 vs tf32 vs fp32.
All was reporting well, except this setup: [t5-large on A100](https://github.com/huggingface/transformers/issues/15026#issuecomment-1004543189) for some reason was completely not activating bf16 optimizations. 

This is with BS=8. Observe:

| Variation       |   Train<br>samples<br>per<br>second |   Diff<br>% |   Train<br>loss |
|:----------------|------------------------------------:|------------:|----------------:|
| --tf32 0        |                               31.59 |         0 |            2.03 |
| --tf32 1        |                               36.13 |         14 |            2.03 |
| --bf16 --tf32 0 |                               31.35 |          -1 |            2.04 |
| --bf16 --tf32 1 |                               31.30 |          -1 |            2.04 |

As it's clear, bf16 is slower than fp32 here.

Here is the same benchmark on RTX-3090

| Variation       |   Train<br>samples<br>per<br>second |   Diff<br>% |   Train<br>loss |
|:----------------|------------------------------------:|------------:|----------------:|
| --tf32 0        |                               31.88 |         0 |            2.03 |
| --tf32 1        |                               35.66 |         12 |            2.03 |
| --bf16 --tf32 0 |                               35.07 |         10 |            2.04 |
| --bf16 --tf32 1 |                               35.13 |         10 |            2.04 |

no problem here, bf16 or tf32 are faster than fp32.

Talking to @eqy on pytorch slack he told me that the batch size was too small. (note the same bs=8 on both cards). 

So I go for the max before OOM, which is bs=16, and we get on A100:

| Variation       |   Train<br>samples<br>per<br>second |   Diff<br>% |   Train<br>loss |
|:----------------|------------------------------------:|------------:|----------------:|
| --tf32 0        |                               39.59 |           0 |            2.10 |
| --tf32 1        |                               67.24 |          70 |            2.10 |
| --bf16 --tf32 0 |                               61.37 |          55 |            2.12 |
| --bf16 --tf32 1 |                               59.95 |          51 |            2.12 |

That's better, but still far from optimal. Only after adding grad accumulation to make an effective BS=64, I was able to get decent outcome:

| Variation       |   Train<br>samples<br>per<br>second |   Diff<br>% |   Train<br>loss |
|:----------------|------------------------------------:|------------:|----------------:|
| --tf32 0        |                               45.60 |           0 |            2.35 |
| --tf32 1        |                               86.78 |          90 |            2.36 |
| --bf16 --tf32 0 |                               75.85 |          66 |            2.41 |
| --bf16 --tf32 1 |                               73.19 |          61 |            2.41 |

(all the details to reproduce the issue are in the benchmark posts: [A100](https://github.com/huggingface/transformers/issues/15026) and [RTX-3090](https://github.com/huggingface/transformers/issues/14608)).

Now how could I possibly know that BS=8 was absolutely bad and BS=16 was still pretty bad and that I needed to squeeze BS=64? 

--------------------

If for example I'm using Megatron-LM it tells the user:
```
    custom_kernel_constraint = seq_len > 16 and seq_len <=2048 and \
        seq_len % 4 == 0 and attn_batch_size % 4 == 0
    # Print a warning.
    if not ((args.fp16 or args.bf16) and
            custom_kernel_constraint and
            args.masked_softmax_fusion):
        if args.rank == 0:
            print('WARNING: constraints for invoking optimized'
                  ' fused softmax kernel are not met. We default'
                  ' back to unfused kernel invocations.', flush=True)
```
if its custom fused kernels aren't going to get activated.

Could we have something similar for CUDA optimizations?

So what I am asking for is a similar feature for CUDA. So if say I run:

```
REPORT_UNMET_CONSTRAINTS=1 python pytorch-program.py 
```

In the case of BS=8 or BS=16 above it'll tell me that some constraints haven't been met and ideally which, so that the user can act upon it.

Thank you.

**Comment by eqy** [comment]
In this case it is more likely that larger batch sizes amortize overheads in the optimizer and non-GEMM parts of the model. It would be great if performance differences could be root-caused to singular constraints, but here the answer is likely to be more complex. For example, if the optimizer is slow relative to the forward/backward pass, this can mask any efficiency improvements from reduced precision. The reason increasing the effective batch size can help is by increasing the ratio of forward/backward time (more efficient w/ optimized GEMM kernels) to optimizer time (not specialized to reduced precision types).

**Comment by stas00** [comment]
But how do you explain that the same setup on RTX-3090 doesn't have that issue? Surely if it's top level software related we should have seen the same relative problem with the other card.

Given that at bs=16 things rectify you don't think it has to do with the magical 16x constraint?

But then this benchmark: Benchmark 5: gpt2-medium uses BS=8 and it doesn't have this problem:


| Variation       |   Train<br>samples<br>per<br>second |   Diff<br>% |   Train<br>loss |
|:----------------|------------------------------------:|------------:|----------------:|
| --tf32 0        |                               10.60 |         0 |            2.98 |
| --tf32 1        |                               24.81 |         134 |            2.98 |
| --fp16 --tf32 0 |                               27.67 |         161 |            2.99 |
| --fp16 --tf32 1 |                               27.62 |         160 |            2.99 |
| --bf16 --tf32 0 |                               27.57 |         160 |            2.99 |
| --bf16 --tf32 1 |                               27.55 |         160 |            2.99 |

so bs=16 is not magical.

> Overall, I think a typical approach in "PyTorch style" is to instrument the training loop of the workload. It may be crude, but a few torch.cuda.synchronize calls and timers would uncover this issue.

For an advanced user it probably would help. But a lay user won't have an idea how to do that or even that their setup is sub-optimal.

But I'm not insisting that there must be a way if you know there is not. I'm just suggesting that perhaps there could be a way. Perhaps we need to talk to NVIDIA and provide a channel to the user for when some constraint isn't met. e.g. like `NCCL_DEBUG=INFO` providing a useful trace of what it is doing if that could be a good example.

**Comment by eqy** [comment]
I'm not sure we would expect the same bottlenecks to materialize in the same way across 3090 and A100 given that they have different ratios of compute (FLOPs) to memory bandwidth. This situation becomes even more complex when we consider TF32, as A100 (with its already higher FLOPs/bandwidth ratio relative to 3090 becomes even more skewed that way).

Due to T5-Large being a bigger model, it looks like `T5LayerNorm` is a bigger bottleneck compared to the smaller T5s. As for the difference with GPT-2, this could also be explained by the use of the `T5LayerNorm` if it is not used by GPT-2. `T5LayerNorm` also introduces additional casts when using reduced precision: https://github.com/huggingface/transformers/blob/37ed3ab719f10dc00bf63ac343b441bf78bb1eee/src/transformers/models/t5/modeling_t5.py#L245 which may not help performance relative to fp32 or tf32.

Thus far, I have been using the profiler/profiling to uncover bottlenecks in the current benchmark. I'm not really qualified to comment on workflows other than the ones I've used, but in my experience PyTorch has encouraged active user involvement in understanding the performance characteristics of workloads (consider the stats for data loading displayed by the widely used ImageNet example code).

**Comment by ngimel** [comment]
Related #55655. cc @bitfort, @robieta, this is related to torch tidy that we are planning.

**Comment by stas00** [comment]
Thank you for the link to the pre-existing proposal, @ngimel!

Yes, very similar to @ezyang's proposal and yes, to be enabled by the user explicitly so that there is no worrying about the possible slowdown for the production use.

Please let me know if I should close this issue if the one you linked to is already covering all the bases. Here I'm just giving more concrete examples that we could use as a test case for when such a tool would flag an issue or a few.

I'd add specifically examples of when crucial tensor dimensions are not divisible by 8 and 16 as an additional fodder for such a tool. Ideally it should recommend to pad such inputs to either 8 or 16 depending on the actual constraint.

**Comment by ngimel** [comment]
Having concrete examples of what performance recommendations would be useful is always great, so thank you for this issue! Let's leave it open for now.
